server/app/validation.py

```diff
@@ -1,92 +1,234 @@
-import enum
```

Comment:
I will skip this file since you are still looking for your style, I guess :)

Reply:
If you have ideas on the validation, I want to hear them!

I like `attrs` better than `pydantic` because it doesn't do implicit parsing and casting and whatnot, where it's difficult to know what is actually happening. However, I'm still searching for options to not have to duplicate validators and converters over and over.

Reply:
Sure, still finding my own style with `attrs` though :D

I would split up the file `validation.py` into a directory `types/` with one file for each purpose.

It is a good rule of thumb that when something is really hard to statically type, then it might be too complicated.

Reply:
> I would split up the file `validation.py` into a directory `types/` with one file for each purpose.

Something like one module for query validation, one for body validation, and one for the combined request validation?

> [...] when something is really hard to statically type, then it might be too complicated.

I have to force myself to type a bit 😄 Some things you have to write just differently.

Do you have an example of this? Or something where you think I could simplify the code to make typing easier?

Reply:
> Something like one module for query validation, one for body validation, and one for the combined request validation?

Maybe. I thought of putting the first few sections ("Route validation base classes and functions", "Attrs converters", "Attrs validators", "Attrs fields") in a file `validation/utils.py`, the route validation into `validation/routes.py` and the mqtt validation into `validation/mqtt.py`.

With a `validation/__init__.py` you can explicitly reexport things so that you have a nice "table of contents of the `validation` directory.

> Do you have an example of this? Or something where you think I could simplify the code to make typing easier?

I haven't worked myself into your code that deep to find those :D. It's just that when I use dynamic Python types, I am not forced to have "clean function interfaces" and can just throw data structures around - which is obviously not good. This is very noticeable when trying to statically type these interfaces. With clean interfaces, the "add types afterwards" approach can be done very quick.

Reply:
> Maybe. I thought of putting [...]

Sounds good, I'll try that 👍